aws_subnet on creation improperly handles IPv6 CIDR pending association state

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

```
Terraform v1.1.7
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.12.1
```
### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* aws_subnet

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
locals { 
    azs =  { "us-west-2a" : 0, "us-west-2b" : 1, "us-west-2c" : 2 }
    vpc_main_cidr = "10.0.0.0/16"
}

resource "aws_vpc" "main" {
  cidr_block                       = local.vpc_main_cidr
  assign_generated_ipv6_cidr_block = true
  enable_dns_support               = true
  enable_dns_hostnames             = true

  tags = {
    Name          = "${terraform.workspace}-main"
  }
}

resource "aws_subnet" "subnet" {
  for_each = local.azs

  vpc_id            = aws_vpc.main.id

  availability_zone = each.key
  cidr_block        = cidrsubnet(aws_vpc.main.cidr_block, 8, each.value)
  ipv6_cidr_block   = cidrsubnet(aws_vpc.main.ipv6_cidr_block, 8, each.value)

  map_public_ip_on_launch         = true #Need an IPv4 public
  assign_ipv6_address_on_creation = true #IPv6 Please Sir

  tags = {
    Name           = "${terraform.workspace}-main-public"
  }
}

```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

This is a dev environment and the output has been modified/redacted 
https://gist.github.com/moses-aronov/3ea2124e5d7a51998a0a5480fa30dc79

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

It should successfully create the subnet. 

### Actual Behavior

There is an intermittent error. 
The subnet fails to fully create. 

```
│ Error: error associating EC2 Subnet (subnet-0162ef609ddae733e) IPv6 CIDR block (2600:1f13:9cd:e501::/64): InvalidSubnet.Conflict: The CIDR '2600:1f13:9cd:e501::/64' conflicts with another subnet
│       status code: 400, request id: 75c51f61-1542-4dd0-b94e-2fca79f8e804
│
│   with aws_subnet.main_public_rgn_3["us-west-2b"],
│   on vpc_us_west_2.tf line 80, in resource "aws_subnet" "main_public_rgn_3":
│   80: resource "aws_subnet" "main_public_rgn_3" {
```

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`
2.  You may get the following error
```
│ Error: error associating EC2 Subnet (subnet-0162ef609ddae733e) IPv6 CIDR block (2600:1f13:9cd:e501::/64): InvalidSubnet.Conflict: The CIDR '2600:1f13:9cd:e501::/64' conflicts with another subnet
│       status code: 400, request id: 75c51f61-1542-4dd0-b94e-2fca79f8e804
│
│   with aws_subnet.main_public_rgn_3["us-west-2b"],
│   on vpc_us_west_2.tf line 80, in resource "aws_subnet" "main_public_rgn_3":
│   80: resource "aws_subnet" "main_public_rgn_3" {
```


### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

If the subnet is available but the IPv6 CIDR block is still associating, we will run into this error.

After we create subnet we wait for the IPv6 CIDR block to become associated. Once associated we return the subnet object without updating the IPv6 cidr association.

https://github.com/hashicorp/terraform-provider-aws/blob/25a5ad33014830744b832d593c63486df5d38dda/internal/service/ec2/vpc_subnet.go#L197

This will cause problems later in the creation process as we check for associated CIDRs

https://github.com/hashicorp/terraform-provider-aws/blob/25a5ad33014830744b832d593c63486df5d38dda/internal/service/ec2/vpc_subnet.go#L405

### References


<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->



* #0000

We have been seeing this error in nightly CI:

```
=== RUN   TestAccEC2SubnetCidrReservation_ipv6
=== PAUSE TestAccEC2SubnetCidrReservation_ipv6
=== CONT  TestAccEC2SubnetCidrReservation_ipv6
vpc_subnet_cidr_reservation_test.go:53: Step 1/2 error: Error running apply: exit status 1
Error: error associating EC2 Subnet (subnet-0b96a077b2c236dc8) IPv6 CIDR block (2600:1f14:8bf:c901::/64): InvalidSubnet.Conflict: The CIDR '2600:1f14:8bf:c901::/64' conflicts with another subnet
  status code: 400, request id: 49620341-403f-412e-940d-c39c4edda2dc
with aws_subnet.test,
on terraform_plugin_test.tf line 11, in resource "aws_subnet" "test":
11: resource "aws_subnet" "test" {
--- FAIL: TestAccEC2SubnetCidrReservation_ipv6 (53.78s)
```

--- a/.changelog/24685.txt
+++ b/.changelog/24685.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_subnet: Fixes aws_subnet on creation improperly handling IPv6 CIDR pending association state #24685
+```

--- a/.changelog/24685.txt
+++ b/.changelog/24685.txt
@@ -1,3 +1,7 @@
 ```release-note:bug
 resource/aws_subnet: Fix `InvalidSubnet.Conflict` errors when associating IPv6 CIDR blocks
 ```
+
+```release-note:bug
+resource/aws_default_subnet: Fix `InvalidSubnet.Conflict` errors when associating IPv6 CIDR blocks
+```

--- a/.changelog/24685.txt
+++ b/.changelog/24685.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-resource/aws_subnet: Fixes aws_subnet on creation improperly handling IPv6 CIDR pending association state #24685
+resource/aws_subnet: Fix `InvalidSubnet.Conflict` errors when associating IPv6 CIDR blocks
 ```

--- a/internal/service/ec2/vpc_default_subnet.go
+++ b/internal/service/ec2/vpc_default_subnet.go
@@ -200,15 +200,17 @@ func resourceDefaultSubnetCreate(d *schema.ResourceData, meta interface{}) error
 		}
 
 		// Creating an IPv6-native default subnets associates an IPv6 CIDR block.
-		for _, v := range subnet.Ipv6CidrBlockAssociationSet {
+		for i, v := range subnet.Ipv6CidrBlockAssociationSet {
 			if aws.StringValue(v.Ipv6CidrBlockState.State) == ec2.SubnetCidrBlockStateCodeAssociating { //we can only ever have 1 IPv6 block associated at once
 				associationID := aws.StringValue(v.AssociationId)
 
-				_, err = WaitSubnetIPv6CIDRBlockAssociationCreated(conn, associationID)
+				subnetCidrBlockState, err := WaitSubnetIPv6CIDRBlockAssociationCreated(conn, associationID)
 
 				if err != nil {
 					return fmt.Errorf("error waiting for EC2 Default Subnet (%s) IPv6 CIDR block (%s) to become associated: %w", d.Id(), associationID, err)
 				}
+
+				subnet.Ipv6CidrBlockAssociationSet[i].Ipv6CidrBlockState = subnetCidrBlockState
 			}
 		}
 

--- a/internal/service/ec2/vpc_subnet.go
+++ b/internal/service/ec2/vpc_subnet.go
@@ -200,7 +200,7 @@ func resourceSubnetCreate(d *schema.ResourceData, meta interface{}) error {
 				return fmt.Errorf("error waiting for EC2 Subnet (%s) IPv6 CIDR block (%s) to become associated: %w", d.Id(), associationID, err)
 			}
 
-			subnet.Ipv6CidrBlockAssociationSet[i].SetIpv6CidrBlockState(subnetCidrBlockState)
+			subnet.Ipv6CidrBlockAssociationSet[i].Ipv6CidrBlockState = subnetCidrBlockState
 		}
 	}
 

--- a/internal/service/ec2/vpc_subnet.go
+++ b/internal/service/ec2/vpc_subnet.go
@@ -190,15 +190,17 @@ func resourceSubnetCreate(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("error waiting for EC2 Subnet (%s) create: %w", d.Id(), err)
 	}
 
-	for _, v := range subnet.Ipv6CidrBlockAssociationSet {
+	for i, v := range subnet.Ipv6CidrBlockAssociationSet {
 		if aws.StringValue(v.Ipv6CidrBlockState.State) == ec2.SubnetCidrBlockStateCodeAssociating { //we can only ever have 1 IPv6 block associated at once
 			associationID := aws.StringValue(v.AssociationId)
 
-			_, err = WaitSubnetIPv6CIDRBlockAssociationCreated(conn, associationID)
+			subnetCidrBlockState, err := WaitSubnetIPv6CIDRBlockAssociationCreated(conn, associationID)
 
 			if err != nil {
 				return fmt.Errorf("error waiting for EC2 Subnet (%s) IPv6 CIDR block (%s) to become associated: %w", d.Id(), associationID, err)
 			}
+
+			subnet.Ipv6CidrBlockAssociationSet[i].SetIpv6CidrBlockState(subnetCidrBlockState)
 		}
 	}
 

